File: frontend/src/types/application.ts
export type ApplicationStatus = 
  | 'saved' 
  | 'applied' 
  | 'assessment' 
  | 'interview' 
  | 'final_interview' 
  | 'offer' 
  | 'rejected' 
  | 'withdrawn';

export interface Application {
  id: string;
  user_id: string;
  job_id: string;
  status: ApplicationStatus;
  applied_date?: string;
  notes?: string;
  resume_used?: string;
  cover_letter_used?: string;
  created_at: string;
  updated_at: string;
  
  // Joined from jobs table
  job_title?: string;
  job_company?: string;
  job_location?: string;
}

export interface ApplicationStats {
  total: number;
  saved: number;
  applied: number;
  assessment: number;
  interview: number;
  final_interview: number;
  offer: number;
  rejected: number;
  withdrawn: number;
}

export interface GeneratedResume {
  id: string;
  template_name: string;
  match_score?: number;
  optimization_notes?: string;
  pdf_url?: string;
  tex_content?: string;
  created_at: string;
}

export interface GeneratedCoverLetter {
  id: string;
  content: string;
  job_id?: string;
  ai_model: string;
  created_at: string;
}

export interface BulletEnhanceRequest {
  bullet_point: string;
  target_role?: string;
  target_skills?: string[];
}

export interface BulletEnhanceResponse {
  original_bullet: string;
  optimized_bullet: string;
  impact_explanation: string;
  alternatives: string[];
}

export interface MasterResumeData {
  has_resume: boolean;
  file_path?: string | null;
  download_url?: string | null;
  parsed?: {
    skills?: string[];
    experience?: Array<{ title?: string; company?: string; start_date?: string; end_date?: string; description?: string }>;
    projects?: Array<{ title?: string; description?: string; technologies?: string[] }>;
    education?: Array<{ degree?: string; institution?: string; field_of_study?: string }>;
    raw_text?: string;
  } | null;
  profile?: {
    name?: string;
    email?: string;
    phone?: string;
    location?: string;
    linkedin?: string;
    github?: string;
    portfolio?: string;
  };
  updated_at?: string;
}

File: backend/app/models/application.py
"""
EngineerCopilot AI — Application & Resume Pydantic Models.

Defines schemas for application tracking, resume generation,
and cover letter generation.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# ── Applications ────────────────────────────────────────────

VALID_STATUSES = [
    "saved", "applied", "assessment", "interview",
    "final_interview", "offer", "rejected", "withdrawn",
]


class ApplicationCreate(BaseModel):
    job_id: str
    status: str = Field(default="saved", pattern=r"^(saved|applied|assessment|interview|final_interview|offer|rejected|withdrawn)$")
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None


class ApplicationUpdate(BaseModel):
    status: Optional[str] = Field(None, pattern=r"^(saved|applied|assessment|interview|final_interview|offer|rejected|withdrawn)$")
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    resume_used: Optional[str] = None
    cover_letter_used: Optional[str] = None


class ApplicationResponse(BaseModel):
    id: str
    user_id: str
    job_id: str
    status: str
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    resume_used: Optional[str] = None
    cover_letter_used: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    # Joined job data (optional)
    job_title: Optional[str] = None
    job_company: Optional[str] = None
    job_location: Optional[str] = None

    class Config:
        from_attributes = True


class ApplicationStats(BaseModel):
    total: int = 0
    saved: int = 0
    applied: int = 0
    assessment: int = 0
    interview: int = 0
    final_interview: int = 0
    offer: int = 0
    rejected: int = 0
    withdrawn: int = 0


# ── Resume Generation ──────────────────────────────────────

class ResumeGenerateRequest(BaseModel):
    job_id: Optional[str] = None
    template_name: str = "ats_classic"
    custom_job_description: Optional[str] = None  # If no job_id, use raw text


class ResumeGenerateResponse(BaseModel):
    id: str
    template_name: str
    match_score: Optional[int] = None
    optimization_notes: Optional[str] = None
    pdf_url: Optional[str] = None
    tex_content: Optional[str] = None
    created_at: datetime


class ResumeParseResponse(BaseModel):
    """Result of resume parsing."""
    skills: list[str] = Field(default_factory=list)
    education: list[dict] = Field(default_factory=list)
    experience: list[dict] = Field(default_factory=list)
    projects: list[dict] = Field(default_factory=list)
    certifications: list[dict] = Field(default_factory=list)
    raw_text: str = ""


# ── Cover Letter ────────────────────────────────────────────

class CoverLetterGenerateRequest(BaseModel):
    job_id: Optional[str] = None
    company_name: Optional[str] = None
    job_title: Optional[str] = None
    job_description: Optional[str] = None


class CoverLetterResponse(BaseModel):
    id: str
    content: str
    job_id: Optional[str] = None
    ai_model: str = "gemini"
    created_at: datetime

    class Config:
        from_attributes = True


# ── Saved Jobs ──────────────────────────────────────────────

class SavedJobResponse(BaseModel):
    id: str
    user_id: str
    job_id: str
    created_at: datetime
    job_title: Optional[str] = None
    job_company: Optional[str] = None
    job_location: Optional[str] = None
    job_source: Optional[str] = None
